### ISSUE TITLE
xcd panicked at Result::unwrap() after calling truss on it

### ISSUE THREAD

**Issue by daniloegea** [michael-yuji/xc#4]
The xcd child process will panic if you try to call `truss` on it.

How to reproduce:

1. Run a container
`xc run -d freebsdxc/freebsd:13.2 sh /etc/rc`

2. Call `truss` on the xcd child process
`truss -p 19706`

3. Watch it panicking

The code is panicking here after kevent unblocks: [xc/src/container/runner/mod.rs#L604](https://github.com/michael-yuji/xc/blob/main/xc/src/container/runner/mod.rs#L604)

The error is `EINTR: Interrupted system call`.

The workaround is to check if `kevent_ts` returned Err and jump back to the start of the loop but I'm still trying to properly understand the full picture here.

Here's the stack trace.

```
2023-10-02T18:13:49.964741Z  INFO xc::container::runner: exited: 0
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: EINTR', xc/src/container/runner/mod.rs:604:28
stack backtrace:
   0:     0x3bc58a0d1441 - std::backtrace_rs::backtrace::libunwind::trace::h51e064ea2aa96e0f
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x3bc58a0d1441 - std::backtrace_rs::backtrace::trace_unsynchronized::h2813d8db85a5db15
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x3bc58a0d1441 - std::sys_common::backtrace::_print_fmt::h261f3ebf1326ac10
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x3bc58a0d1441 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc84b874768c0bb91
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x3bc58a10183f - core::fmt::rt::Argument::fmt::hcf2b2643095ccb42
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/rt.rs:138:9
   5:     0x3bc58a10183f - core::fmt::write::h16a408d2d9c17d71
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/mod.rs:1094:21
   6:     0x3bc58a0cc867 - std::io::Write::write_fmt::h8e4e50496a5c1e88
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/io/mod.rs:1714:15
   7:     0x3bc58a0d1245 - std::sys_common::backtrace::_print::h5f15d18412853039
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x3bc58a0d1245 - std::sys_common::backtrace::print::h9eb58656c53ff807
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x3bc58a0d2b63 - std::panicking::default_hook::{{closure}}::hac8b9301429e8db0
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:269:22
  10:     0x3bc58a0d284a - std::panicking::default_hook::hf27d88dc07654b00
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:288:9
  11:     0x3bc58a0d31dd - std::panicking::rust_panic_with_hook::h7b8285c824ce3974
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:705:13
  12:     0x3bc58a0d2fe7 - std::panicking::begin_panic_handler::{{closure}}::h01be5fefe7a5a158
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:597:13
  13:     0x3bc58a0d18a6 - std::sys_common::backtrace::__rust_end_short_backtrace::h6d40530c867dd7f3
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:151:18
  14:     0x3bc58a0d2d32 - rust_begin_unwind
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:593:5
  15:     0x3bc589241773 - core::panicking::panic_fmt::h75a44213c16c1672
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panicking.rs:67:14
  16:     0x3bc589241d03 - core::result::unwrap_failed::he2a594c760b21776
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/result.rs:1651:5
  17:     0x3bc5896c0e8b - core::result::Result<T,E>::unwrap::h69fe214cd9b7b76f
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/result.rs:1076:23
  18:     0x3bc589796896 - xc::container::runner::ProcessRunner::run::h786b41db8064e3eb
                               at /usr/home/danilo/Sources/xc/xc/src/container/runner/mod.rs:604:23
  19:     0x3bc58979be9d - xc::container::runner::run::h8bdcf04d1d08c3fe
                               at /usr/home/danilo/Sources/xc/xc/src/container/runner/mod.rs:743:17
  20:     0x3bc589369403 - xcd::site::Site::run_container::hcce8d5af634823e5
                               at /usr/home/danilo/Sources/xc/xcd/src/site.rs:466:21
  21:     0x3bc58946d81f - xcd::context::ServerContext::continue_instantiate::{{closure}}::h375bdc4bc7250780
                               at /usr/home/danilo/Sources/xc/xcd/src/context.rs:572:13
  22:     0x3bc5894733f5 - xcd::context::ServerContext::instantiate::{{closure}}::h28c06e49cc780ef8
                               at /usr/home/danilo/Sources/xc/xcd/src/context.rs:655:65
  23:     0x3bc58931e7d1 - <xcd::ipc::__ImplStructinstantiate as ipc::service::Method<<alloc::sync::Arc<tokio::sync::rwlock::RwLock<xcd::context::ServerContext>> as ipc::util::ExtractInner>::Inner,<ipc::service::ConnectionContext<xcd::ipc::Variables> as ipc::util::ExtractInner>::Inner>>::apply::{{closure}}::inner::{{closure}}::h2d7f594540a843af
                               at /usr/home/danilo/Sources/xc/xcd/src/ipc.rs:338:22
  24:     0x3bc58931d5bc - <xcd::ipc::__ImplStructinstantiate as ipc::service::Method<<alloc::sync::Arc<tokio::sync::rwlock::RwLock<xcd::context::ServerContext>> as ipc::util::ExtractInner>::Inner,<ipc::service::ConnectionContext<xcd::ipc::Variables> as ipc::util::ExtractInner>::Inner>>::apply::{{closure}}::h047ded896a56a9ab
                               at /usr/home/danilo/Sources/xc/xcd/src/ipc.rs:309:1
  25:     0x3bc58927f2ae - <core::pin::Pin<P> as core::future::future::Future>::poll::h00b0a101f1d875ed
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/future/future.rs:125:9
  26:     0x3bc58926e3c2 - ipc::service::Stream<T,V>::inner::{{closure}}::h9fbfaf915fb88f30
                               at /usr/home/danilo/Sources/xc/ipc/src/service/mod.rs:151:74
  27:     0x3bc58926ead1 - ipc::service::Stream<T,V>::activate::{{closure}}::h104c2f94b313f263
                               at /usr/home/danilo/Sources/xc/ipc/src/service/mod.rs:174:46
  28:     0x3bc589270283 - ipc::service::Service<T,V>::accept::{{closure}}::{{closure}}::h95c9e06d1eaa5d58
                               at /usr/home/danilo/Sources/xc/ipc/src/service/mod.rs:219:31
  29:     0x3bc58956d9a4 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::h6d721ee8838919f1
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/core.rs:334:17
  30:     0x3bc58956a26f - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h860cf34cac86e78c
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/loom/std/unsafe_cell.rs:16:9
  31:     0x3bc58956a26f - tokio::runtime::task::core::Core<T,S>::poll::h022e35decccbbecb
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/core.rs:323:13
  32:     0x3bc58952d435 - tokio::runtime::task::harness::poll_future::{{closure}}::hef3c287c18938f03
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:485:19
  33:     0x3bc5895868e4 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h4898e40626a217e4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panic/unwind_safe.rs:271:9
  34:     0x3bc58951cd9c - std::panicking::try::do_call::h87e4ac40672d8c5c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:500:40
  35:     0x3bc58952590b - __rust_try
  36:     0x3bc589516728 - std::panicking::try::h73cb1abfccb1285c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:464:19
  37:     0x3bc58938359b - std::panic::catch_unwind::hd703305d3cb6ff69
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panic.rs:142:14
  38:     0x3bc589528c9f - tokio::runtime::task::harness::poll_future::h895202f370c668ea
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:473:18
  39:     0x3bc5895302c9 - tokio::runtime::task::harness::Harness<T,S>::poll_inner::hfc53062d50e19168
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:208:27
  40:     0x3bc589535487 - tokio::runtime::task::harness::Harness<T,S>::poll::hd054f8c23f6070da
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:153:15
  41:     0x3bc58958bf5d - tokio::runtime::task::raw::poll::h95521ba050e186fc
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/raw.rs:276:5
  42:     0x3bc589f6dd27 - tokio::runtime::task::raw::RawTask::poll::hc69cb279d842e8f6
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/raw.rs:200:18
  43:     0x3bc589f97992 - tokio::runtime::task::LocalNotified<S>::run::h3a5a98944bfb110a
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/mod.rs:400:9
  44:     0x3bc589fed7cd - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}::h4c768313d58686e7
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:576:13
  45:     0x3bc589fed617 - tokio::runtime::coop::with_budget::hcc675e819306c435
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:107:5
  46:     0x3bc589fed617 - tokio::runtime::coop::budget::h3cad86362f46ed79
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:73:5
  47:     0x3bc589fed617 - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h06ac468ed617e536
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:575:9
  48:     0x3bc589fecbef - tokio::runtime::scheduler::multi_thread::worker::Context::run::h724197d9392a7d6f
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:526:24
  49:     0x3bc589fec649 - tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::{{closure}}::h457dc36fb6bbd725
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:491:21
  50:     0x3bc589f85330 - tokio::runtime::context::scoped::Scoped<T>::set::hbf67a22d08aef0bb
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/scoped.rs:40:9
  51:     0x3bc589fbbd3b - tokio::runtime::context::set_scheduler::{{closure}}::hbf9f82b9d1dd0585
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context.rs:176:26
  52:     0x3bc589fb779e - std::thread::local::LocalKey<T>::try_with::h348de1c1b28d1f4d
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/thread/local.rs:270:16
  53:     0x3bc589fb742b - std::thread::local::LocalKey<T>::with::h6102315d8efc6bb3
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/thread/local.rs:246:9
  54:     0x3bc589fbbc74 - tokio::runtime::context::set_scheduler::h0d021bc2ccdcc1f1
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context.rs:176:9
  55:     0x3bc589fec551 - tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::h87d4f70a48e0d464
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:486:9
  56:     0x3bc589ff4d08 - tokio::runtime::context::runtime::enter_runtime::h69c776fb9e5b726e
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/runtime.rs:65:16
  57:     0x3bc589fec2dc - tokio::runtime::scheduler::multi_thread::worker::run::hff1cb0f87867259e
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:478:5
  58:     0x3bc589fec14b - tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}::hfc743b4dde99acb2
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/worker.rs:447:45
  59:     0x3bc589fb6dee - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h740b1bdfc0726952
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/blocking/task.rs:42:21
  60:     0x3bc589f7a4e3 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::h6bb7042c8ddc43e5
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/core.rs:334:17
  61:     0x3bc589f7a1db - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h052a816a4aaed3b7
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/loom/std/unsafe_cell.rs:16:9
  62:     0x3bc589f7a1db - tokio::runtime::task::core::Core<T,S>::poll::haa42bc2b6f847d99
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/core.rs:323:13
  63:     0x3bc589fd7785 - tokio::runtime::task::harness::poll_future::{{closure}}::h7638b249118701fa
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:485:19
  64:     0x3bc589f91813 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hfac52084700d4221
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panic/unwind_safe.rs:271:9
  65:     0x3bc589f9dadc - std::panicking::try::do_call::h0ad60ca5a27c3d14
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:500:40
  66:     0x3bc589f9e9cb - __rust_try
  67:     0x3bc589f9d938 - std::panicking::try::hbf0b1f51dd8ef7fd
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:464:19
  68:     0x3bc589fe493b - std::panic::catch_unwind::h3e278f276f97607a
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panic.rs:142:14
  69:     0x3bc589fd731f - tokio::runtime::task::harness::poll_future::h3d3c575120bf5015
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:473:18
  70:     0x3bc589fd6289 - tokio::runtime::task::harness::Harness<T,S>::poll_inner::hdd60bec4e8a1a669
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:208:27
  71:     0x3bc589fd6007 - tokio::runtime::task::harness::Harness<T,S>::poll::h2f1a4f60a2f08b02
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/harness.rs:153:15
  72:     0x3bc589f6dfad - tokio::runtime::task::raw::poll::hd7d15257e564c9f7
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/raw.rs:276:5
  73:     0x3bc589f6dd27 - tokio::runtime::task::raw::RawTask::poll::hc69cb279d842e8f6
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/raw.rs:200:18
  74:     0x3bc589f97a57 - tokio::runtime::task::UnownedTask<S>::run::hef52159e48dc213c
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/task/mod.rs:437:9
  75:     0x3bc589f97df7 - tokio::runtime::blocking::pool::Task::run::hf1983f8c834a7219
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/blocking/pool.rs:159:9
  76:     0x3bc589f9a449 - tokio::runtime::blocking::pool::Inner::run::hbcd6429d25185556
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/blocking/pool.rs:513:17
  77:     0x3bc589f9a174 - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::h812d5f1e706a351b
                               at /usr/home/danilo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/blocking/pool.rs:471:13
  78:     0x3bc589f73349 - std::sys_common::backtrace::__rust_begin_short_backtrace::ha110b2e017b2ec88
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:135:18
  79:     0x3bc589f68152 - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h123e6e14d6b22ed9
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/thread/mod.rs:529:17
  80:     0x3bc589f91652 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h25930294c3fb1496
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panic/unwind_safe.rs:271:9
  81:     0x3bc589f9db8b - std::panicking::try::do_call::h0f196e6acce0f93d
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:500:40
  82:     0x3bc589f9e9cb - __rust_try
  83:     0x3bc589f9d031 - std::panicking::try::h10fa9667d5cf95b3
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:464:19
  84:     0x3bc589f67f62 - std::panic::catch_unwind::h06c09309aeccb65f
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panic.rs:142:14
  85:     0x3bc589f67f62 - std::thread::Builder::spawn_unchecked_::{{closure}}::hbda1c694a93373d1
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/thread/mod.rs:528:30
  86:     0x3bc589fd7dbf - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc8681363445ecda0
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/ops/function.rs:250:5
  87:     0x3bc58a0d8d55 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he6b6ab83bd1ed5f5
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/alloc/src/boxed.rs:1993:9
  88:     0x3bc58a0d8d55 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc738cb5b60b285bd
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/alloc/src/boxed.rs:1993:9
  89:     0x3bc58a0d8d55 - std::sys::unix::thread::Thread::new::thread_start::hcf79b46ddedf30d4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys/unix/thread.rs:108:17
  90:     0x3bcdaf928a75 - <unknown>

```

**Comment by daniloegea** [comment]
Another issue is that it's not collecting the children status after exit and leaving zombies behind.

**Comment by Freaky** [comment]
[Note](https://docs.rs/nix/latest/x86_64-unknown-freebsd/nix/sys/event/fn.kevent_ts.html): "Deprecated since 0.27.0: Use [Kqueue::kevent](https://docs.rs/nix/latest/x86_64-unknown-freebsd/nix/sys/event/struct.Kqueue.html#method.kevent) instead"

The `Err` should be checked against `Err(nix::errno::Errno::EINTR)` and the loop retried (or perhaps more safely, the kevent call placed in its own loop just to check for EINTR).

**Comment by khng300** [comment]
Calling truss on process then causing the system call in-progress returning EINTR is an expected behaviour. Just to re-enter the loop instead of crashing the child process is fine.

**Comment by daniloegea** [comment]
Thank you for your comments, I created a PR with a workaround https://github.com/michael-yuji/xc/pull/5

The recovering process still needs to be improved and we probably should move to the new nix crate version (it will require some refactoring).
